### PR TITLE
Batch build smaller provider packages using `up xpkg batch`

### PR DESCRIPTION
**Pull request by ulucinar**
### Description of your changes

Depends on: https://github.com/upbound/up/pull/332

This PR invokes the `up xpkg batch` command to streamline the processing of the smaller provider packages. The `xpkg build` implementation dumps the built packages onto the file system for `xpkg push` to load them back from the filesystem and push them to a package registry. The `xpkg batch` command does not need to store the built provider packages in the filesystem. The packages are instead directly pushed to the configured registry. 

All smaller packages built by `xpkg batch` share a common base image and the `img.build` make target is modified so that it now builds only this common base image instead of the individual smaller provider images. `xpkg batch` adds the `base` and the `upbound` layers as before, as well as a layer containing the provider controller binary. The `base` and `upbound` labeled layers are shared between all the architectures of a provider's package.

`xpkg batch` can also build & push the provider packages in parallel with the configured (via the `--concurrency` option) maximum concurrency. 

`xpkg batch` also has a templating engine so that the family of smaller providers being built can share a template provider package metadata. The templating engine substitutes the `{{ .Service }}` and `{{ .Name }}` template variables with the service and package repo names, respectively, and further template variables can also be supplied via multiple `--template-var` command-line options. This template metadata resides at `package/crossplane.yaml.tmpl` and is introduced with this PR. 

This PR also bumpt `up` to `v0.17.0`.

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

TODO: 
- [ ] We will need to adjust uptest tests as they rely on the provider package being dumped onto the filesystem.

### How has this code been tested

Tested manually by replacing the `up` invoked by the new `batch-process` make target with an `up` binary compiled from the https://github.com/upbound/up/pull/332 PR.

### CODE REVIEW

File: package/crossplane.yaml.tmpl [hunk 1]
```diff
@@ -1,24 +1,28 @@
 apiVersion: meta.pkg.crossplane.io/v1alpha1
 kind: Provider
 metadata:
-  name: provider-gcp-{{ APIGROUP }}
+  name: {{ .Name }}
+{{ if ne .Service "monolith" }}
   labels:
-    pkg.crossplane.io/provider-family: gcp
+    pkg.crossplane.io/provider-family: provider-family-gcp
```
Comment (turkenh) [inline package/crossplane.yaml.tmpl:7]: nit: Should we flow the `CSP` makefile variable down to here?

File: package/crossplane.yaml.tmpl [hunk 2]
```diff
@@ -1,24 +1,28 @@
...
-    friendly-name.meta.crossplane.io: Provider GCP ({{ APIGROUP }})
+      address. This is a subpackage for the {{ .Service }} API group.
+    friendly-name.meta.crossplane.io: Provider GCP ({{ .Service }})
+    auth.upbound.io/group: gcp.upbound.io
+{{ if and (ne .Service "config") (ne .Service "monolith") -}}
 spec:
   dependsOn:
-    - provider: {{ XPKG_REG_ORGS }}/provider-gcp-config
-      version: "{{ DEP_CONSTRAINT }}"
+    - provider: {{ .XpkgRegOrg }}/provider-family-gcp
```
Comment (turkenh) [inline package/crossplane.yaml.tmpl:26]: If this what we finally aligned in terms of naming the config package?

**Review by turkenh — APPROVED** [review]
LGTM, thanks @ulucinar!

**Review by turkenh — APPROVED** [review]
LGTM, thanks @ulucinar!

File: Makefile [hunk 3]
```diff
@@ -1,9 +1,11 @@
 # ====================================================================================
 # Setup Project
 
-PROJECT_NAME := provider-gcp
+CSP := gcp
```
[the remark this answers is not in the document]
Comment (ulucinar) [reply Makefile:4]: It's an acronym for [cloud service provider](https://cloud.google.com/learn/what-is-a-cloud-service-provider). Let's replace it with `PROVIDER_NAME`.

File: Makefile [hunk 4]
```diff
@@ -88,8 +77,24 @@ export UP_CHANNEL := $(UP_CHANNEL)
```
[the remark this answers is not in the document]
Comment (ulucinar) [reply Makefile:77]: Done.

[the remark this answers is not in the document]
Comment (ulucinar) [reply Makefile:77]: And thank you very much @turkenh for preparing the release!

File: Makefile [hunk 5]
```diff
@@ -88,8 +77,24 @@ export UP_CHANNEL := $(UP_CHANNEL)
 
 REGISTRY_ORGS ?= xpkg.upbound.io/upbound
 IMAGES = provider-gcp
+BATCH_PLATFORMS ?= linux_amd64,linux_arm64
+export BATCH_PLATFORMS := $(BATCH_PLATFORMS)
```
[the remark this answers is not in the document]
Comment (ulucinar) [reply Makefile:81]: The main make module receives a `PLATFORMS` variable with value `linux_amd64 linux_arm64`. This is what we expect except for the delimiter. But during publishing, the image make module receives it as `darwin_amd64 darwin_arm64 windows_amd64 linux_amd64 linux_arm64`. I did not inspect further. Let's revisit this later.

File: package/crossplane.yaml.tmpl (hunk 1 above)
Comment (ulucinar) [reply package/crossplane.yaml.tmpl]: Done.

File: package/crossplane.yaml.tmpl (hunk 2 above)
Comment (ulucinar) [reply package/crossplane.yaml.tmpl]: Done. I renamed the family package from`provider-family-gcp` to `provider-gcp-config`, which is consistent with our smaller provider naming scheme.

**Pull request merged**